File: Python/Daily Quest/02-medium/NumberOfWaysToAssignEdgeWeightsI.py
""" There is an undirected tree with n nodes labeled from 1 to n, rooted at node 1. The tree is represented by a 2D integer array edges of length n - 1, where edges[i] = [ui, vi] indicates that there is an edge between nodes ui and vi.

Initially, all edges have a weight of 0. You must assign each edge a weight of either 1 or 2.

The cost of a path between any two nodes u and v is the total weight of all edges in the path connecting them.

Select any one node x at the maximum depth. Return the number of ways to assign edge weights in the path from node 1 to x such that its total cost is odd.

Since the answer may be large, return it modulo 109 + 7.

Note: Ignore all edges not in the path from node 1 to x. """

class Solution:
    def assignEdgeWeights(self, edges: list[list[int]]) -> int:
        grafo = {}

        for u, v in edges:
            if u not in grafo:
                grafo[u] = []

            grafo[u].append(v)

            if v not in grafo:
                grafo[v] = []

            grafo[v].append(u)

        max_arestas = 0

        def recursive(no_atual, no_pai, edgesAccumulated):
            nonlocal max_arestas
            max_arestas = max(max_arestas, edgesAccumulated)

            for vizinho in grafo[no_atual]:
                if (vizinho != no_pai):
                    recursive(vizinho, no_atual, edgesAccumulated + 1)

        recursive(1, 0, 0)
        
        return pow(2, max_arestas-1, 10**9 + 7)
    

File: Cpp/medium/NumberOfWaysToAssignEdgeWeightsI.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    long long fast_pow(long long base, long long expoente, long long modulo)
    {
        long long resultado = 1;
        base %= modulo;
        while (expoente > 0)
        {
            if (expoente % 2 == 1)
            {
                resultado = (resultado * base) % modulo;
            }
            base = (base * base) % modulo;
            expoente /= 2;
        }
        return resultado;
    }

    void dfs(int no_atual, int no_pai, int edgesAccumulated, const unordered_map<int, vector<int>> &grafo, int &max_edges)
    {
        max_edges = std::max(max_edges, edgesAccumulated);

        for (auto vizinho : grafo.at(no_atual))
        {
            if (vizinho != no_pai)
            {
                dfs(vizinho, no_atual, edgesAccumulated + 1, grafo, max_edges);
            }
        }
    }

    int assignEdgeWeights(vector<vector<int>>& edges)
    {
        unordered_map<int, vector<int>> grafo = {};

        for (auto u : edges)
        {
            grafo[u[0]].push_back(u[1]);
            grafo[u[1]].push_back(u[0]);
        }

        int max_edges = 0;

        dfs(1, 0, 0, grafo, max_edges);

        return fast_pow(2, max_edges - 1, 1e9 + 7);
    }
};